Source/DafnyLanguageServer/Language/Symbols/DafnyLangSymbolResolver.cs

```diff
@@ -150,9 +154,43 @@ private class SymbolDeclarationResolver {
           cancellationToken.ThrowIfCancellationRequested();
           functionSymbol.Parameters.Add(ProcessFormal(scope, parameter));
         }
+        if (function.Result != null) {
+          functionSymbol.Parameters.Add(ProcessFormal(scope, function.Result));
+        }
+        ScopeSymbol? ExpressionHandler(Expression? expression) =>
+          expression == null
+            ? null
+            : ProcessExpression(new ScopeSymbol(functionSymbol, functionSymbol.Declaration), expression);
+        functionSymbol.Body = ExpressionHandler(function.Body);
+        functionSymbol.Ens = ProcessListAttributedExpressions(function.Ens, ExpressionHandler);
+        functionSymbol.Req = ProcessListAttributedExpressions(function.Req, ExpressionHandler);
+        functionSymbol.Reads = ProcessListFramedExpressions(function.Reads, ExpressionHandler);
+        functionSymbol.Decreases = ProcessListExpressions(function.Decreases.Expressions, ExpressionHandler);
+        functionSymbol.ByMethodBody = function.ByMethodBody == null ? null : ProcessFunctionByMethodBody(functionSymbol, function.ByMethodBody);
         return functionSymbol;
       }
 
+      private List<ScopeSymbol> ProcessListAttributedExpressions(
+        IList<AttributedExpression> list, Func<Expression?, ScopeSymbol?> expressionHandler
+        ) {
+        return list.SelectMany(attributedExpression =>
+          expressionHandler(attributedExpression.E).AsEnumerable()).ToList();
+      }
+
+      private List<ScopeSymbol> ProcessListFramedExpressions(
+        IList<FrameExpression> list, Func<Expression?, ScopeSymbol?> expressionHandler
+        ) {
+        return list.SelectMany(frameExpression =>
+          expressionHandler(frameExpression.E).AsEnumerable()).ToList();
```

Comment:
Thanks for the explanation for the use of `AsEnumerable`.
But I think it's easier to understand if you add a `Where(frameExpression => frameExpression.E != null)` filter.

Reply:
It would be a little better for the human reader, however, the incomplete type checker will complain as

    List<ScopeSymbol?> x = ...;
    x .Where(x => x != null)

the last expression is also of type `List<ScopeSymbol?>` and not `List<ScopeSymbol>`.

Reply:
You may either use the null-forgiving operator `!` and/or create a suitable extension method. For example:
```cs
    public static IEnumerable<TElement> WhereNotNull<TElement>(this IEnumerable<TElement?> elements) where TElement : class {
      foreach(var element in elements) {
        if(element != null) {
          yield return element;
        }
      }
    }
```

Or within the LINQ-expression using the null-forgiving operator:
```cs
return list
  .Where(frameExpression => frameExpression.E != null)
  .Select(frameExpression => expressionHandler(frameExpression.E!));
```